crates/matrix-sdk-base/src/store/mod.rs

```diff
@@ -276,115 +277,184 @@ impl Deref for Store {
-/// moment, and basically does the job.
-#[derive(Debug)]
-struct ObservableMap<K, V>
-where
-    V: Clone + Send + Sync + 'static,
-{
-    /// The (key, position) tuples.
-    mapping: HashMap<K, usize>,
-
-    /// The values where the indices are the `position` part of `Self::mapping`.
-    values: ObservableVector<V>,
-}
+#[cfg(not(target_arch = "wasm32"))]
+mod observable_map {
+    use std::{borrow::Borrow, collections::HashMap, hash::Hash};
 
-impl<K, V> ObservableMap<K, V>
-where
-    K: Hash + Eq,
-    V: Clone + Send + Sync + 'static,
-{
-    /// Create a new `Self`.
-    fn new() -> Self {
-        Self { mapping: HashMap::new(), values: ObservableVector::new() }
-    }
+    use eyeball_im::{ObservableVector, Vector, VectorDiff};
+    use futures_util::Stream;
 
-    /// Insert a new `V` in the collection.
+    /// An observable map.
+    ///
+    /// This is an “observable map” naive implementation. Just like regular
+    /// hashmap, we have a redirection from a key to a position, and from a
+    /// position to a value. The (key, position) tuples are stored in an
+    /// [`HashMap`]. The (position, value) tuples are stored in an
+    /// [`ObservableVector`]. The (key, position) tuple is only provided for
+    /// fast _reading_ implementations, like `Self::get` and
+    /// `Self::get_or_create`. The (position, value) tuples are observable,
+    /// this is what interests us the most here.
     ///
-    /// The position of the inserted value is returned.
+    /// Why not implementing a new `ObservableMap` type in `eyeball-im` instead
+    /// of this custom implementation? Because we want to continue providing
+    /// `VectorDiff` when observing the changes, so that the rest of the API in
+    /// the Matrix Rust SDK aren't broken. Indeed, an `ObservableMap` must
+    /// produce `MapDiff`, which would be quite different.
+    /// Plus, we would like to re-use all our existing code, test, stream
+    /// adapters and so on.
     ///
-    /// If the `V` value already exists, it will be updated to the new one.
-    fn insert(&mut self, key: K, value: V) -> usize {
-        match self.mapping.get(&key) {
-            Some(position) => {
-                self.values.set(*position, value);
+    /// This is a trade-off. This implementation is simple enough for the
+    /// moment, and basically does the job.
+    #[derive(Debug)]
+    pub(super) struct ObservableMap<K, V>
+    where
+        V: Clone + Send + Sync + 'static,
+    {
+        /// The (key, position) tuples.
+        mapping: HashMap<K, usize>,
 
-                *position
-            }
-            None => {
-                let position = self.values.len();
+        /// The values where the indices are the `position` part of
+        /// `Self::mapping`.
+        values: ObservableVector<V>,
+    }
+
+    impl<K, V> ObservableMap<K, V>
+    where
+        K: Hash + Eq,
+        V: Clone + Send + Sync + 'static,
```

Comment:
Do we need the `Clone` bound here?

Reply:
Yes, because of `eyeball_im::ObservableVector`, https://docs.rs/eyeball-im/latest/eyeball_im/struct.ObservableVector.html#impl-ObservableVector%3CT%3E.